manifests/source.pp

```diff
@@ -80,177 +101,269 @@
+        }
+      } elsif $key {
+        if $key =~ Hash {
+          unless $key['name'] or $key['id'] {
+            fail('key hash must contain a key name (for apt::keyring) or an id (for apt::key)')
+          }
+          if $key['id'] {
+            # defaults like keyserver are only relevant to apt::key
+            $_key = $apt::source_key_defaults + $key
+          } else {
+            $_key = $key
+          }
+        } else {
+          $_key = { 'id' => assert_type(String[1], $key) }
+        }
+        if $_key['ensure'] {
+          $_key_ensure = $_key['ensure']
+        } else {
+          $_key_ensure = $ensure
+        }
+
+        # Old keyserver keys handled by apt-key
+        if $_key =~ Hash and $_key['id'] {
+          # We do not want to remove keys when the source is absent.
+          if $ensure == 'present' {
+            apt::key { "Add key: ${$_key['id']} from Apt::Source ${title}":
+              ensure   => $_key_ensure,
+              id       => $_key['id'],
+              server   => $_key['server'],
+              content  => $_key['content'],
+              source   => $_key['source'],
+              options  => $_key['options'],
+              weak_ssl => $_key['weak_ssl'],
+              before   => $_before,
+            }
+          }
+          $_list_keyring = undef
+        }
+        # Modern apt keyrings
+        elsif $_key =~ Hash and $_key['name'] {
+          apt::keyring { $_key['name']:
+            ensure   => $_key_ensure,
+            content  => $_key['content'],
+            source   => $_key['source'],
+            dir      => $_key['dir'],
+            filename => $_key['filename'],
+            mode     => $_key['mode'],
+            before   => $_before,
+          }
+
+          $_list_keyring = if $_key['dir'] and $_key['filename'] {
+            "${_key['dir']}${_key['filename']}"
+          } elsif $_key['filename'] {
+            "/etc/apt/keyrings/${_key['filename']}"
+          } elsif $_key['dir'] {
+            "${_key['dir']}${_key['name']}"
+          } else {
+            "/etc/apt/keyrings/${_key['name']}"
+          }
+        }
+      } else {
+        # No `key` nor `keyring` provided
+        $_list_keyring = undef
       }
-      if $key['id'] {
-        # defaults like keyserver are only relevant to apt::key
-        $_key = $apt::source_key_defaults + $key
+
+      $header = epp('apt/_header.epp')
+
+      if $architecture {
+        $_architecture = regsubst($architecture, '\baarch64\b', 'arm64')
       } else {
-        $_key = $key
+        $_architecture = undef
+      }
+
+      $source_content = epp('apt/source.list.epp', {
+          'comment'          => $comment,
+          'includes'         => $includes,
+          'options'          => delete_undef_values({
+              'arch'              => $_architecture,
+              'trusted'           => $allow_unsigned ? { true => 'yes', false => undef, default => undef },
```

Comment:
This feel wrong as `allow_unsigned => true` leads to `trusted => 'yes'`,

I am understanding this wrongly, or did you mean:
```suggestion
              'trusted'           => $allow_unsigned ? { false => 'yes', default => undef },
```

Reply:
This is largely the same as what's in main currently: https://github.com/puppetlabs/puppetlabs-apt/blob/b565ce0b6d203eabb122d87e62db4e5457a6f4bc/manifests/source.pp#L219

I think it's correct as-is. 
`trusted => true` essentially results in "I trust this repo, do not fail even if the repo is unsigned".
and `trusted => false` results in no addition within the generated sources.list file, which by default means the source is not "trusted".

Reply:
Ah I see, thanks!  I updated #1212 which is based on your work to improve templates and tests accordingly.